fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,463 @@
+        let cache = self.height_data.read()?;
+        Ok(cache.get_value(height).map(|i| i.1.clone()))
+    }
+
+    fn all_top_down_msgs(&self) -> StmResult<Vec<CrossMsg>> {
+        let cache = self.top_down_msgs.read()?;
+        Ok(cache.values().cloned().collect())
+    }
+}
+
+#[async_trait]
+impl ParentViewProvider for DefaultFinalityProvider {
+    fn latest_height(&self) -> StmDynResult<Option<BlockHeight>> {
+        let h = self.parent_view_data.latest_height()?;
+        Ok(h)
+    }
+
+    fn latest_nonce(&self) -> StmDynResult<Option<Nonce>> {
+        let top_down_msgs = self.parent_view_data.top_down_msgs.read()?;
+        Ok(top_down_msgs.upper_bound())
+    }
+
+    fn new_block_height(
+        &self,
+        height: BlockHeight,
+        block_hash: BlockHash,
+        validator_set: ValidatorSet,
+    ) -> StmDynResult<()> {
+        let insert_res = self.parent_view_data.height_data.modify(|mut cache| {
+            let r = cache.append(height, (block_hash.clone(), validator_set.clone()));
+            (cache, r)
+        })?;
+
+        match insert_res {
+            Ok(_) => Ok(()),
+            Err(_) => {
+                // now the inserted height is not the next expected block height, could be a chain
+                // reorg if the caller is behaving correctly.
+                abort(Error::ParentReorgDetected(height))
+            }
+        }
+    }
+
+    fn new_top_down_msgs(&self, top_down_msgs: Vec<CrossMsg>) -> StmDynResult<()> {
+        if top_down_msgs.is_empty() {
+            // not processing if there are no top down msgs
+            return Ok(());
+        }
+
+        ensure_sequential_by_nonce(&top_down_msgs)?;
+
+        // get the min nonce from the list of top down msgs and purge all the msgs with nonce
+        // about the min nonce in cache, as the data should be newer and more accurate.
+        let min_nonce = top_down_msgs.first().unwrap().msg.nonce;
+        self.parent_view_data.top_down_msgs.update(|mut cache| {
+            cache.remove_key_above(min_nonce);
+
+            for msg in top_down_msgs.clone() {
+                // safe to unwrap, as the append is sequential
+                cache.append(msg.msg.nonce, msg).unwrap();
+            }
+            cache
+        })?;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl ParentFinalityProvider for DefaultFinalityProvider {
+    fn last_committed_finality(&self) -> StmDynResult<IPCParentFinality> {
+        let finality = self.last_committed_finality.read_clone()?;
+        Ok(finality)
+    }
+
+    fn next_proposal(&self) -> StmDynResult<IPCParentFinality> {
+        let latest_height = if let Some(h) = self.parent_view_data.latest_height()? {
+            h
+        } else {
+            return abort(Error::HeightNotReady);
+        };
+
+        // latest height has not reached, we should wait or abort
+        if latest_height < self.config.chain_head_delay {
```

Comment:
This check only applies to the `(0, chain_head_delay)` interval, i.e. the first few blocks?

Reply:
yep, just the first few.